programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+    {
+        return( MBEDTLS_ERR_SERIALIZE_UNSUPPORTED_OUTPUT );
+    }
+    header[0] = MBEDTLS_SERIALIZE_TYPE_RESULT;
+    header[1] = length >> 16 & 0xff;
+    header[2] = length >> 8 & 0xff;
+    header[3] = length & 0xff;
+    if( (ret = mbedtls_serialize_write( ctx, header, sizeof( header ) ) ) != 0 )
+        return( ret );
+    return( mbedtls_serialize_write( ctx, buffer, length ) );
+}
+
+/** Read one message from the serialization channel and process it.
+ *
+ * For a push message, push the input parameter onto the stack.
+ * For an execute message, execute the function and send the results.
+ *
+ * If the channel is dead (as indicated by ctx->status), do nothing.
+ * If ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY, ignore
+ * parameters and reply MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED to the next
+ * function then set the status back to OK.
+ * In case of any I/O error, set ctx->status to
+ * MBEDTLS_SERIALIZE_STATUS_DEAD. */
+static int mbedtls_serialize_pull( mbedtls_serialize_context_t *ctx )
+{
+    int ret;
+    uint8_t header[4];
+    if( ctx->status == MBEDTLS_SERIALIZE_STATUS_DEAD )
+    {
+        DBG( "already dead" );
+        return( MBEDTLS_ERR_SERIALIZE_RECEIVE );
+    }
+    if( ( ret = mbedtls_serialize_read( ctx, header, sizeof( header ) ) ) != 0 )
+    {
+        DBG( "receive failure -> dead" );
+        ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+        return( ret );
+    }
+    switch( header[0] )
+    {
+    case MBEDTLS_SERIALIZE_TYPE_PUSH:
+    {
+        size_t length = header[1] << 16 | header[2] << 8 | header[3];
+        mbedtls_serialize_item_t *item;
+        DBG( "received push length=%zu", length );
+        if( ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY )
+        {
+            /* While we're out of memory, keep reading arguments but discard
+               them. */
+            while( length > sizeof( header ) )
+            {
+                if( ( ret = mbedtls_serialize_read( ctx, header, sizeof( header ) ) ) != 0 )
+                {
+                    DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                    ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                    return( ret );
+                }
+                length -= sizeof( header );
+            }
+            if( ( ret = mbedtls_serialize_read( ctx, header, length ) ) != 0 )
+            {
+                DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                return( ret );
+            }
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
+        }
+        item = alloc_item( length );
+        if( item == NULL )
+        {
+            DBG( "failed to allocate %zu bytes for input", length );
+            ctx->status = MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY;
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
```

Comment:
Could you elaborate on the intended behavior in case of not enough memory? The first time memory for an item cannot be allocated, the `ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY` branch is not taken (because so far things have been ok), and when `alloc_item` fails afterwards, `ctx->status` will be set to `MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY` and the function immediately fails without dumping the buffer content. So at the end of this first failing call, the input to `mbedtls_serialize_read` is the opaque buffer content, which on the next invocation of `mbedtls_serialize_pull` will be interpreted as a `header` instance.

Shouldn't the dumping of the item contents be done after checking that `item == NULL` and before the function returns, so that `mbedtls_serialize_pull` always consumes exactly one full message, regardless of whether it could allocate memory for it or not?